lib/main.js

```diff
@@ -0,0 +1,153 @@
+/*!
+ * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
+ */
+import * as assert from './assert.js';
+import noopLogger from './noopLogger';
+import {
+  UnauthorizedClient, InvalidGrant, InvalidRequest, InvalidClient
+} from '@interop/oauth2-errors';
+
+// Default Access token expiration in seconds (two weeks)
+const DEFAULT_MAX_AGE = 1209600;
+
+/**
+ * Constructs a middleware handler for OAuth 2.0 /token exchange requests.
+ *
+ * @param {object} options - Options hashmap.
+ * @param {Function} options.authenticateClient - Async callback that
+ *   authenticates the client token.
+ * @param {object} [options.defaults={}] - Token issuance defaults.
+ * @param {object} [options.logger] - Logger.
+ * @param {Function} options.getClient - Async callback to load client by ID.
+ * @param {Function} options.issue - Async callback that issues the JWT access
+ *   token for the client.
+ *
+ * @example <caption>Example usage.</caption>
+ * tokenExchangeHandler({
+ *   defaults: {},
+ *   logger: console,
+ *   getClient: async ({clientId}) => {
+ *     // load client from the db, then:
+ *     return {client};
+ *   }),
+ *   issue: async ({client, scope, body, defaultMaxAge}) => {
+ *     // issue JWT access token here
+ *     return {accessToken, expiresIn};
+ *   })
+  },
+ * });
+ *
+ * @returns {Promise<Function>} Resolves with exchange middleware.
+ */
+function tokenExchangeHandler({
+  authenticateClient,
+  defaults = {},
+  getClient,
+  issue,
+  logger = noopLogger,
+}) {
+  assert.isFunction(authenticateClient, 'authenticateClient');
+  assert.isFunction(getClient, 'getClient');
+  assert.isFunction(issue, 'issue');
+
+  return async (req, res) => {
+    try {
+      const {body} = req;
+      const {
+        client_id: clientId,
+        client_secret: clientSecret,
+        grant_type: grantType,
+        scope: scopeRequested
+      } = body;
+
+      _validateParams({clientId, clientSecret, grantType});
+
+      let client;
+      try {
+        client = await getClient({clientId});
+      } catch(cause) {
+        const error = new InvalidRequest({
+          description: 'Could not load client.'
+        });
+        error.cause = cause;
+        throw error;
+      }
+
+      const {authenticated} = await authenticateClient({client, clientSecret});
+      if(!authenticated) {
+        throw new InvalidClient({
+          description: 'The client secret could not be authenticated.'
+        });
+      }
+
+      const defaultMaxAge = defaults.defaultMaxAge || DEFAULT_MAX_AGE;
```

Comment:
Pardon that this is not related to the PR, but there is an issue here if zero is a valid value for defaultMaxAge.

Reply:
@davidlehn thank you for double-checking.
You're right (which I suppose is why the nullish coalescing operator can't arrive soon enough).
In this case though, zero is not a valid value (it'll mean that the tokens immediately expire).

Reply:
@dmitrizagidulin That landed in Node.js 14 so you can use it today.  Should probably put an engine field in package.json if you do so.  Use in browsers might require some more consideration to make sure tooling can handle it.

Reply:
@davidlehn Oh, good to know! Thanks!